tests/utils.py

```diff
@@ -141,26 +136,45 @@ class TestServerProcess():
   """
 
 
-  def __init__(self, log, server='simple_server.py',
-      port=None, timeout=10, popen_cwd=".",
-      extra_cmd_args=[]):
+  def __init__(self, log, server='simple_server.py', timeout=10,
+      popen_cwd=".", extra_cmd_args=[]):
 
     # Create temporary log file used for logging stdout and stderr
     # of the subprocess. In the mode "r+"" stands for reading and writing
     # and "t" stands for text mode.
     self.__temp_log_file = tempfile.TemporaryFile(mode='r+t')
 
     self.server = server
-    self.port = port or random.randint(30000, 45000)
+    self.port = random.randint(30000, 45000)
     self.__logger = log
 
-    # The "-u" option forces stdin, stdout and stderr to be unbuffered.
-    command = ['python', '-u', server, str(self.port)] + extra_cmd_args
-
-    # We are reusing one server subprocess in multiple unit tests, but we are
-    # collecting the logs per test.
-    self.__server_process = subprocess.Popen(command,
-        stdout=self.__temp_log_file, stderr=subprocess.STDOUT, cwd=popen_cwd)
+    started = False
+    while not started:
```

Comment:
this loop needs to handle timeout somehow: if `has_server_started()` keeps saying no, this loop never exits.

it might make more sense to keep both loops in this function for that reason (the timeout being shared), and instead have the separate function just be the check for logfile contents and poll()

Reply:
Great discovery!